### DISCUSSION TITLE
Rest API Auth Question (Possibility / Strategy)

### DISCUSSION THREAD

**Discussion by VikingDadMedic** [Budibase/budibase#5468] [Help]
There is a 3rd party rest api I need to connect with to query some analytics data. I am having difficulty properly constructing the request (see Connection Requirements below). 

Is constructing a 'per request' _base-64 encoded HMAC-SHA256 signature_ possible using Budibase? And **if so, can someone post an example or high level overview of how this is accomplished inside Budibase?** 

If not possibly in straight forward method - workaround for Budibase? Could a secondary API be used to generate this request ([example](https://documenter.getpostman.com/view/3362843/RWMCt9WU)) or using automation in process, or running some JS in the UI per button click (i.e. request execution)? Can [CryptoJS](https://cryptojs.gitbook.io/docs/#hmac) be pulled into the process?

**Connection Requirements** _(per 3rd party docs)_

> The API is validated using both a public key and a private HMAC Hash of the request contents. These values must be passed in the request header as `X-Public-Key` and `X-Signed-Request-Hash`.
> 
> It is a base-64 encoded HMAC-SHA256 signature. 
> 
> **The signature will need to be computed for each request**. You can generate `SECRETHASH` using an HMAC-SHA256 hash of the request **TYPE** _(Ex. POST)_, **URL** _(ex. user/234256/loginLink)_ and any **REQUEST DATA** _(ex. { 'plan_id': 34 })_ using `APISECRET` as the key, and base-64 encoding the result. 

Related: #2685 #3230

**Comment by melohagan** [comment]
Hi @VikingDadMedic 

I'm not sure how feasible this is because you can't use external libraries in your JavaScript. 

Otherwise you would have been able to generate the signature and pass it through to the headers of your request using bindings: 
 - https://docs.budibase.com/docs/rest-bindings#headers-example
 - https://docs.budibase.com/docs/custom-queries#using-query-bindings

If you're feeling adventurous, you could try adding a 'Generate Hash' action to the codebase that uses CryptoJS, but you can also raise a feature request. 

Using an 'Execute Query' action that grabs the signature from an external API, and passes it onto your second 'Execute Query' request could potentially work: https://docs.budibase.com/docs/referencing

**Comment by RainardSmith** [comment]
Hi @VikingDadMedic 
Have you got any luck with this? I'm trying to achieve the same.

Thanks!!
